src/component/legend/LegendView.js

```diff
@@ -191,10 +191,15 @@ export default echarts.extendComponentView({
                     }
 
                     if (seriesModel.legendDataProvider) {
-                        var data = seriesModel.legendDataProvider();
+                        // pie/funnel legend can use visualMap color.
+						var data = seriesModel.getData();
                         var idx = data.indexOfName(name);
                         if (idx < 0) {
-                            return;
+							var rawData = seriesModel.legendDataProvider();
+                            idx = rawData.indexOfName(name);
+                            if (idx < 0) {
+                                return;
+                            }
                         }
```

Comment:
I'm not sure about this. Will this leak the implementation of `legend` component to the series?

Reply:
>  leak the implementation of legend component to the series

I think it probably OK about that, now that the name is `legendDataProvider`.
The original way is: 
series give a "data" to legend, and then legend retrieve what 
it wants from the "data". 
But now the problem we meet is that legend can not retrieve some 
attr from the data (e.g., can not retrieve `color` and `borderColor`). So series
need to provide more ways for legend to make sure it can retrieve them. That 
probably has made the fact that: series knows that legend need `color` and `borderColor`.

Reply:
I think we can modify the `getVisualByName` to
```js
getVisualByName: function (name, visualProperty) {
        var data = series.getData();
        var idx = data.indexOfName(name);
        return data.getItemVisual(name, visualProperty);
}
```

Reply:
@pissang I used to consider this way. But this way make `data.indexOfName(name)` execute twice if both retrieve `color` and `borderColor`. `data.indexOfName(name)` cost `O(log(n))`. Is it OK. Or may be cache can be used inside `data.indexOfName(name)` if performance needs to take into account.

Reply:
This extra performance cost is acceptable in the `legend` scene. Which data amount is usually small.

Reply:
Now I've no idea about how to choose. Which one do you prefer?

Reply:
I prefer the way with better performance :)

BTW: I think TypeScript will help a lot in this scene.

Reply:
OK, we can use the last solution.
```js
var index = provider.indexOfName(name);
var color = provider.getVisual(index, 'color');
var borderColor = provider.getVisual(index, 'borderColor');
```